src/main/kotlin/org/jitsi/rtp/rtp/RedPacketParser.kt

```diff
@@ -0,0 +1,255 @@
+import org.jitsi.rtp.util.getBitsAsInt
+import java.lang.IllegalArgumentException
+import kotlin.experimental.and
+import kotlin.experimental.or
+
+/**
+ * Parses RED (RFC2198) packets.
+ */
+class RedPacketParser<PacketType : RtpPacket>(
+    /**
+     * A function to create packets from redundancy blocks (used so we can create a packet with the correct class).
+     */
+    val createPacket: (ByteArray, Int, Int) -> PacketType
+) {
+    /**
+     * Destructively parses a specific [RtpPacket] as RED. The redundant RED blocks are removed from the payload,
+     * leaving only the primary block as payload, and the payload type is changed to the payload type of the primary
+     * block. The transformed packet is no longer a RED packet, so the operation must not be performed again.
+     *
+     * If [parseRedundancy] is set, the redundancy blocks are parsed as [PacketType] into newly allocated buffers.
+     *
+     * @param rtpPacket the packet to parse
+     * @param parseRedundancy whether to parse redundancy packets
+     * @return The list of parsed redundancy packets.
+     */
+    fun decapsulate(
+        rtpPacket: RtpPacket,
+        parseRedundancy: Boolean
+    ): List<PacketType> = with(rtpPacket) {
+        var currentOffset = payloadOffset
+
+        val redundancyBlockHeaders = mutableListOf<RedundancyBlockHeader>()
+        lateinit var primaryBlockHeader: PrimaryBlockHeader
+        do {
+            if (currentOffset > length) {
+                throw IllegalArgumentException(
+                    "Invalid RED packet: no last header block found within the allowed length.")
+            }
+            val blockHeader = BlockHeader.parse(buffer, currentOffset)
+            when (blockHeader) {
+                is PrimaryBlockHeader -> primaryBlockHeader = blockHeader
+                is RedundancyBlockHeader -> redundancyBlockHeaders.add(blockHeader)
+            }
+            currentOffset += blockHeader.headerLength
+        } while (blockHeader is RedundancyBlockHeader)
+
+        val packets = if (parseRedundancy) mutableListOf<PacketType>() else null
+        for ((index, it) in redundancyBlockHeaders.withIndex()) {
+            val blockLength = it.length
+            if (currentOffset + blockLength > offset + length) {
+                throw IllegalArgumentException("Invalid RED packet: blocks extend past the packet length.")
+            }
+
+            if (parseRedundancy) {
+                val byteArray = BufferPool.getArray(
+                    BYTES_TO_LEAVE_AT_START_OF_PACKET +
+                        FIXED_HEADER_SIZE_BYTES + blockLength +
+                        BYTES_TO_LEAVE_AT_END_OF_PACKET
+                )
+
+                System.arraycopy(
+                    buffer, offset,
+                    byteArray, BYTES_TO_LEAVE_AT_START_OF_PACKET,
+                    FIXED_HEADER_SIZE_BYTES
+                )
+                RtpHeader.setCsrcCount(byteArray, BYTES_TO_LEAVE_AT_START_OF_PACKET, 0)
```

Comment:
Just paranoia - does the right thing happen if the incoming RED packet contains CSRCs?

Reply:
I think we're safe -- I just copy the fixed header, so the recovered packets have no CSRCs. I now see that 2198 recommends that we assume the recovered packets inherit the CSRCs from the RED packet. I don't think this makes a difference in our case.